objects/src/mock/chain.rs

```diff
@@ -55,3 +55,401 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+            pending.nullifiers.clear(); // nullifiers are saved in the nullifier TSTM
+        }
+
+        /// Creates a [SimpleSmt] tree from the `notes`.
+        ///
+        /// The root of this tree is a commitment to [Note]s included in the block. The commitment
+        /// is not for the fields in the [Note] structure, but for the [NoteMetadata] and for a
+        /// subset of the [Note]'s attributes (i.e. recipient + vault).
+        pub fn build_notes_tree(&self) -> Result<SimpleSmt, MerkleError> {
+            let mut entries = Vec::with_capacity(self.notes.len() * 2);
+
+            entries.extend(self.notes.iter().enumerate().map(|(index, note)| {
+                let tree_index = (index * 2) as u64;
+                (tree_index, note.hash().into())
+            }));
+            entries.extend(self.notes.iter().enumerate().map(|(index, note)| {
+                let tree_index = (index * 2 + 1) as u64;
+                (tree_index, note.metadata().into())
+            }));
+
+            SimpleSmt::with_leaves(NOTE_LEAF_DEPTH, entries)
+        }
+
+        /// Given the [BlockHeader] and its notedb's [SimpleSmt], set all the [Note]'s proof.
+        ///
+        /// Update the [Note]'s proof once the [BlockHeader] has been created.
+        fn set_notes_proofs(&mut self, header: BlockHeader, notes: &SimpleSmt) {
+            self.notes.iter_mut().enumerate().for_each(|(index, note)| {
+                let auth_index =
+                    NodeIndex::new(NOTE_TREE_DEPTH, index as u64).expect("index bigger than 2**20");
+                let note_path =
+                    notes.get_path(auth_index).expect("auth_index outside of SimpleSmt range");
+                note.set_proof(
+                    NoteInclusionProof::new(
+                        header.block_num(),
+                        header.sub_hash(),
+                        header.note_root(),
+                        index as u64,
+                        note_path,
+                    )
+                    .expect("Invalid data provided to proof constructor"),
+                );
+            });
+        }
+    }
+
+    /// Structure chain data, used to build necessary openings and to construct [BlockHeader]s.
+    #[derive(Debug, Clone)]
+    #[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+    pub struct MockChain<R: Rng + SeedableRng> {
+        /// An append-only structure used to represent the history of blocks produced for this chain.
+        chain: ChainMmr,
+
+        /// Storage for the historical state of the chain.
+        ///
+        /// This store contains historical data for accounts, nullifiers, notes, and batches.
+        data: DefaultMerkleStore,
+
+        /// History of produced blocks.
+        blocks: Vec<BlockHeader>,
+
+        /// Tree containing the latest `Nullifier`'s tree.
+        nullifiers: TieredSmt,
+
+        /// Tree containing the latest hash of each account.
+        accounts: SimpleSmt,
```

Comment:
Maybe leave at `TODO` here to indicate that this should be updated in the future to a Tiered SMT with 64-bit keys.

Reply:
added